internal/controller/compute/user/user.go

```diff
@@ -0,0 +1,339 @@
+
+	for _, groupID := range cr.Spec.ForProvider.GroupIDs {
+		_, _, gerr := eu.service.AddUserToGroup(ctx, groupID, utils.GetStringValue(observed.GetId()))
+		// skip if user is already member of this group.
+		if gerr != nil && strings.Contains(gerr.Error(), "is already member of") {
+			continue
+		}
+		if gerr != nil {
+			return managed.ExternalCreation{ConnectionDetails: conn}, errors.Wrap(err, errAddUserToGroup)
+		}
+	}
+
+	setStatus(cr, observed, cr.Spec.ForProvider.GroupIDs)
+
+	return managed.ExternalCreation{ConnectionDetails: conn}, nil
+}
+
+func (eu *externalUser) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.User)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.Wrap(errors.New(errNotUser), "update error")
+	}
+
+	userID := cr.Status.AtProvider.UserID
+
+	observed, resp, err := eu.service.UpdateUser(ctx, userID, cr.Spec.ForProvider)
+	if err != nil {
+		werr := errors.Wrap(err, errUserUpdate)
+		return managed.ExternalUpdate{}, compute.AddAPIResponseInfo(resp, werr)
+	}
+	conn := connectionDetails(cr, observed)
+
+	err = updateGroups(ctx, eu, userID, cr.Status.AtProvider.GroupIDs, cr.Spec.ForProvider.GroupIDs)
+	if err != nil {
+		return managed.ExternalUpdate{ConnectionDetails: conn}, errors.Wrap(err, "failed to update groups")
+	}
+
+	return managed.ExternalUpdate{ConnectionDetails: conn}, nil
+}
+
+func (eu *externalUser) Delete(ctx context.Context, mg resource.Managed) error {
+	user, ok := mg.(*v1alpha1.User)
+	if !ok {
+		return errors.Wrap(errors.New(errNotUser), "delete error")
+	}
+	user.SetConditions(xpv1.Deleting())
+
+	userID := user.Status.AtProvider.UserID
+
+	for _, groupID := range user.Status.AtProvider.GroupIDs {
+		if err := eu.service.DeleteUserFromGroup(ctx, groupID, userID); err != nil {
+			return errors.Wrap(err, "failed to remove user from a group")
+		}
+	}
+
+	resp, err := eu.service.DeleteUser(ctx, userID)
+	return compute.ErrorUnlessNotFound(resp, errors.Wrap(err, errUserDelete))
+}
+
+// updateGroups adds or remove groups for the userID.
+func updateGroups(ctx context.Context, eu *externalUser, userID string, atProviderGroups []string, forProviderGroups []string) error {
+	for _, groupID := range atProviderGroups {
+		if groupID != "" && !slices.Contains(forProviderGroups, groupID) {
+			if derr := eu.service.DeleteUserFromGroup(ctx, groupID, userID); derr != nil {
+				return errors.Wrap(derr, "failed to remove user from a group")
+			}
+		}
+	}
+
+	for _, groupID := range forProviderGroups {
+		_, _, gerr := eu.service.AddUserToGroup(ctx, groupID, userID)
+		// skip if user is already member of this group.
+		if gerr != nil && strings.Contains(gerr.Error(), "is already member of") {
+			continue
+		}
+		if gerr != nil {
+			return errors.Wrap(gerr, errAddUserToGroup)
+		}
+	}
+	return nil
+}
+
+func createMap(s []string) map[string]bool {
+	m := make(map[string]bool)
+	for i := range s {
+		m[s[i]] = false
+	}
+	return m
+}
```

Comment:
I'd rather use the `slices.Sort + slices.Equal` alternative, but if you want to use this one then it's good to note that it's unecessary to allocate boolean values to this map, a set can be allocated in go using a map[T]interface{}.

```suggestion
func createSet(s []string) map[string]interface{} {
	m := make(map[string]interface{})
	
	for i := range s {
		m[s[i]] = nil
	}
	
	return m
}
```